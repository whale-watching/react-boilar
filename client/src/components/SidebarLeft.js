import React from 'react';
import PropTypes from 'prop-types';
import Close from 'react-icons/lib/md/close';
import Add from 'react-icons/lib/md/add-box';
import Join from 'react-icons/lib/fa/sign-in';
import Leave from 'react-icons/lib/fa/sign-out';
import Lock from 'react-icons/lib/md/lock';
import AvatarModal from './AvatarModal';
import CreateRoomModal from './CreateRoomModal';
import PasswordModal from './PasswordModal';
import { socketEmit } from '../helpers/socketEvents';
import { sidebarClose } from '../helpers/sidebarToggle';

class SidebarLeft extends React.Component {
  constructor() {
    super();

    this.state = {
      avatarModalOpen: false,
      createRoomModalOpen: false,
      passwordModal: {
        open: false,
        roomName: null,
      },
    };

    this.toggleAvatarModal = () => {
      this.setState(prevState => ({ avatarModalOpen: !prevState.avatarModalOpen }));
    };

    this.toggleCreateRoomModal = () => {
      this.setState(prevState => ({ createRoomModalOpen: !prevState.createRoomModalOpen }));
    };

    this.togglePasswordModal = (roomName) => {
      this.setState(prevState => ({
        passwordModal: {
          open: !prevState.passwordModal.open,
          roomName,
        },
      }));
    };

    this.joinRoom = (data) => {
      socketEmit.joinRoom(data.roomName, data.password || null);
    };

    this.leaveRoom = (roomName) => {
      socketEmit.leaveRoom(roomName);
    };

    this.closeSidebar = () => {
      sidebarClose('left');
    };
  }

  render() {
    return (
      <div className="sidebar-left sidebar-left-closed">
        <div className="profile">
          <div className="left">
            <div>
              <img src={this.props.user.avatar} alt="avatar" />
            </div>
            <div>
              <p>{this.props.user && this.props.user.name}</p>
              <button className="edit button-text" title="Edit profile" onClick={this.toggleAvatarModal}>Edit</button>
            </div>
          </div>
          <div className="close">
            <button onClick={this.closeSidebar}>
              <Close className="icon" size="24px" />
            </button>
          </div>
        </div>
        <div className="public-chats">
          <div className="panel">
            <p>Public Chats</p>
            <button onClick={this.toggleCreateRoomModal} title="Create new room">
              <Add className="icon" size="20px" />
            </button>
          </div>
          <div className="public-chats-list">
            {this.props.rooms.map(room => (
              <div className="list">
                <img src={room.name === 'Home Chat' ? '/img/home_chat.png' : '/img/public_chat.png'} alt="chat-room" />
                <div>
                  <p className="primary" onClick={() => this.props.user.rooms.includes(room.name) && this.props.switchRoom(room.name)} title="Switch to this room">{room.name}</p>
                  <p className="secondary">{room.users.join(', ').length > 25 ? `${room.users.join(', ').slice(0, 25)}...` : room.users.join(', ')}</p>
                </div>
                {room.name !== 'Home Chat' && !room.password && (!this.props.user.rooms.includes(room.name) ?
                  <button onClick={() => this.joinRoom({ roomName: room.name })} title="Join this room">
                    <Join className="icon" size="20px" />
                  </button>
                  :
                  <button onClick={() => this.leaveRoom(room.name)} title="Leave this room">
                    <Leave className="icon" size="20px" />
                  </button>)
                }
                {room.password && (!this.props.user.rooms.includes(room.name) ?
                  <button onClick={() => this.togglePasswordModal(room.name)} title="Enter password to join this room">
                    <Lock className="icon" size="20px" />
                  </button>
                  :
                  <button onClick={() => this.leaveRoom(room.name)} title="Leave this room">
                    <Leave className="icon" size="20px" />
                  </button>)
                }
              </div>
            ))}
          </div>
        </div>
        <div className="users">
          <div className="panel">
            <p>Users Online</p>
          </div>
          <div className="users-list">
            {this.props.users.map(user => (
              <div className="list">
                <img src={user.avatar} alt="user-avatar" />
                <div>
                  <p className="primary">{user.name}</p>
                  <p className="secondary">{user.rooms.join(', ').length > 25 ? `${user.rooms.join(', ').slice(0, 25)}...` : user.rooms.join(', ')}</p>
                </div>
              </div>
            ))}
          </div>
        </div>
        <AvatarModal
          isOpen={this.state.avatarModalOpen}
          onRequestClose={this.toggleAvatarModal}
          user={this.props.user}
        />
        <CreateRoomModal
          isOpen={this.state.createRoomModalOpen}
          onRequestClose={this.toggleCreateRoomModal}
        />
        <PasswordModal
          isOpen={this.state.passwordModal.open}
          onRequestClose={() => this.togglePasswordModal(this.state.passwordModal.roomName)}
          roomName={this.state.passwordModal.roomName}
        />
      </div>
    );
  }
}

SidebarLeft.propTypes = {
  user: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    avatar: PropTypes.string,
    rooms: PropTypes.array,
  }).isRequired,
  users: PropTypes.arrayOf(PropTypes.object).isRequired,
  rooms: PropTypes.arrayOf(PropTypes.object).isRequired,
  switchRoom: PropTypes.func.isRequired,
};

export default SidebarLeft;
