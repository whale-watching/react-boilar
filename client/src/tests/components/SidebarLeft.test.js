import React from 'react';
import { shallow } from 'enzyme';
import SidebarLeft from '../../components/SidebarLeft';
import user from '../data/user';
import rooms from '../data/rooms';

describe('<SidebarLeft />', () => {
  it('renders SidebarLeft correctly', () => {
    const wrapper = shallow(<SidebarLeft user={user} users={[user]} rooms={rooms} switchRoom={() => { }} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#avatar').prop('src')).toEqual(user.avatar);
    expect(wrapper.find('#user').text()).toEqual(user.name);
    expect(wrapper.find('.public-chats-list').children()).toHaveLength(rooms.length);
  });
});