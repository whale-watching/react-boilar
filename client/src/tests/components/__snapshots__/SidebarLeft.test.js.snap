// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidebarLeft /> renders SidebarLeft correctly 1`] = `
<div
  className="sidebar-left sidebar-left-closed"
>
  <div
    className="profile"
  >
    <div
      className="left"
    >
      <div>
        <img
          alt="avatar"
          id="avatar"
          src="/img/avatars/1.png"
        />
      </div>
      <div>
        <p
          id="user"
        >
          John
        </p>
        <button
          className="edit button-text"
          onClick={[Function]}
          title="Edit profile"
        >
          Edit
        </button>
      </div>
    </div>
    <div
      className="close"
    >
      <button
        onClick={[Function]}
      >
        <MdClose
          className="icon"
          size="24px"
        />
      </button>
    </div>
  </div>
  <div
    className="public-chats"
  >
    <div
      className="panel"
    >
      <p>
        Public Chats
      </p>
      <button
        onClick={[Function]}
        title="Create new room"
      >
        <MdAddBox
          className="icon"
          size="20px"
        />
      </button>
    </div>
    <div
      className="public-chats-list"
    >
      <div
        className="list"
      >
        <img
          alt="chat-room"
          src="/img/home_chat.png"
        />
        <div>
          <p
            className="primary"
            onClick={[Function]}
            title="Switch to this room"
          >
            Home Chat
          </p>
          <p
            className="secondary"
          >
            John
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="users"
  >
    <div
      className="panel"
    >
      <p>
        Users Online
      </p>
    </div>
    <div
      className="users-list"
    >
      <div
        className="list"
      >
        <img
          alt="user-avatar"
          src="/img/avatars/1.png"
        />
        <div>
          <p
            className="primary"
          >
            John
          </p>
          <p
            className="secondary"
          >
            Home Chat
          </p>
        </div>
      </div>
    </div>
  </div>
  <AvatarModal
    isOpen={false}
    onRequestClose={[Function]}
    user={
      Object {
        "avatar": "/img/avatars/1.png",
        "id": "123",
        "name": "John",
        "rooms": Array [
          "Home Chat",
        ],
      }
    }
  />
  <CreateRoomModal
    isOpen={false}
    onRequestClose={[Function]}
  />
  <PasswordModal
    isOpen={false}
    onRequestClose={[Function]}
    roomName={null}
  />
</div>
`;
